### PR TITLE
Perturbing parameters in nested sub-applications with STM

### LINKED ISSUE #21233: Improve stochastic parameter perturbation for sub apps
## Reason

This issue involves using the stochastic tools module to perturb parameters in a model that has subapps. Currently, the only way to perturb these parameters is to set global parameter in the model's main app which is then passed down using the `cli_args` parameter in the multiapp block. This degrades the model agnostic design of the STM since users will need to set up extra global arguments and use the `cli_args` parameter. And more importantly, this requires the use of `MultiAppSamplerControl` which cannot be used in `batch_restore` mode for more efficient sampling.

## Design

So far I have only thought of input syntax. Say we have a model that has three levels of multiapps, with the last one having multiple using positions:

```
[main] ---- [sub] ---- [subsub0]
              |-------[subsub1]
```

And there is a parameter in `main`: `param0`, a parameter in `sub`: `param1`, a common param in `subsub`: `param2`, and an extra parameter in `subsub0` and `subsub1`: `param3` and `param4`. Here's a rundown:

```
[main  ] ------- [sub   ] -------- [subsub0]
[param0]         [param1]          [param2 ]
                    |              [param3 ]
                    |
                     ------------- [subsub1]
                                   [param2 ]
                                   [param4 ]    
```

For the `MultiAppSamplerControl` the syntax is pretty straightforward:

```
[Controls]
  [param]
    type = MultiAppSamplerControl
    multi_app = main
    sampler = sample
    param_names = 'param0 sub:param1 sub:subsub:param2 sub:subsub0:param3 sub:subsub1:param4'
  []
[]
```

The `SamplerParameterTransfer` is a little more complicated since there needs to be a `SamplerReceiver` at each input. For this, I think the best option to have multiple transfers:

```
[Transfers]
  [param_main]
    type = SamplerParameterTransfer
    sampler = sample
    to_multiapp = main
    to_control = main_receiver
    parameters = 'param0[0]'
  []
  [param_sub]
    type = SamplerParameterTransfer
    sampler = sample
    to_multiapp = main
    to_control = sub:sub_receiver
    parameters = 'param1[1]'
  []
  [param_subsub]
    type = SamplerParameterTransfer
    sampler = sample
    to_multiapp = main
    to_control = sub:subsub:subsub_receiver
    parameters = 'param2[2]'
  []
  [param_subsub0]
    type = SamplerParameterTransfer
    sampler = sample
    to_multiapp = main
    to_control = sub:subsub0:subsub_receiver
    parameters = 'param3[3]'
  []
  [param_subsub1]
    type = SamplerParameterTransfer
    sampler = sample
    to_multiapp = main
    to_control = sub:subsub1:subsub_receiver
    parameters = 'param4[4]'
  []
[]
```

This is just an initial thought. Looking for suggestions.

I'm not sure there's a way to get a subapp's subapp. But this would be required for the implementation.

## Impact

Make STM easier and faster.

### LINKED ISSUE #22827: Improving user experience with stochastic tools module
## Reason

These are some suggestions from users to improve the experience when doing stochastic simulations with the stochastic tools module.

## Design

- [ ] Outputting sampler data along side quantity of interest results
- [ ] Make `Problem/verbose_multiapps=true` show relevant information when in batch mode
- [ ] Show status of sub-app sampling across processors
- [ ] Change file base for batched sub-apps in order to output to different files

## Impact

Better STM experience.

### PR DESCRIPTION
**Pull request by zachmprince**
The main purpose of this PR is the ability to perturb parameters in nested sub-applications like this:

```
[STM app] -> [Main app] -> [Sub App] -> [Subsub App 0]
                                        [Subsub App 1]
```

Which uses the syntax `'<main_param>:<param> sub:<sub_param> sub:subsub:<subsub param> sub:subsub0:<subsub0 param> sub:subsub1:<subsub1 param>'`. This works with MultiAppSamplerControl and SamplerParameterTransfer.

Some other things that are done with this PR:

- Regular command-line arguments for nested multiapps didn't support this syntax, so that was added.
- Changed to using `stringifyExact` for passing command-line arguments with sample multiapp. This is a no-brainer, but required a bunch of regolding.
- SamplerParameterTransfer no longer needs the `to_control` argument since we can just programatically look for the control.

Closes #21233
Closes #22827

### CODE REVIEW

**Review by grmnptr — CHANGES_REQUESTED** [review]
This is definitely very useful addition! I have a few questions though:
1. Since we already have a cascading transfer system to transfer data from subsubsubapps to subsubapps to subapp to the main app, wouldn't it be logical to do something similar for the transfer of parameters too? Or the other way around: if we are already pushing parameters to subsubsubapps directly, shouldn't we pull results directly as well? 
2. I don't understand most of the regolded tests. Is that because of the order of the parameters changed? If not, why did the results change? Did we change seeds? The change is very low but still, I wouldn't expect the modifications to influence old results. 

Also, I am not an expert on framework design so I'll tag @loganharbour as well here.

**Comment by GiudGiud** [comment]
> we are already pushing parameters to subsubsubapps directly

are we doing that now? I must have missed something

File: modules/stochastic_tools/src/transfers/SamplerParameterTransfer.C
Comment (grmnptr) [inline modules/stochastic_tools/src/transfers/SamplerParameterTransfer.C:90]: @GiudGiud I think this gets the parameter receivers in the subapps recursively so it sends the new values directly.

Comment (grmnptr) [reply modules/stochastic_tools/src/transfers/SamplerParameterTransfer.C:90]: As far as I understand, for cases which don't need the controllable parameters, we can use the command line with the `:` separators for subapps.

Comment (GiudGiud) [reply modules/stochastic_tools/src/transfers/SamplerParameterTransfer.C:90]: oh ok it s the whole purpose of the PR

Comment (GiudGiud) [reply modules/stochastic_tools/src/transfers/SamplerParameterTransfer.C:90]: I think it's fine (have not reviewed), we probably should talk about it to get in synch though

**Comment by grmnptr** [comment]
Okay, I read the description of the PR and I think my second question is answered. Just in case:  Did `stringify` crop the parameter samples?

**Comment by zachmprince** [comment]
> 1. Since we already have a cascading transfer system to transfer data from subsubsubapps to subsubapps to subapp to the main app, wouldn't it be logical to do something similar for the transfer of parameters too? Or the other way around: if we are already pushing parameters to subsubsubapps directly, shouldn't we pull results directly as well?

I wasn't able to figure out a way to follow the current cascading approach for `SamplerParameterTransfer` since we aren't transferring global data like solutions or postprocessors, we are basically telling a control what to do later. I would be open to any ideas to do this that follow the same input syntax. Generalizing this capability to all transfers is also something I couldn't figure out, mainly due to the complexities of the "positions" capability. I think @GiudGiud talked about this before, but I forgot the details of the conversation.

> 2. I don't understand most of the regolded tests. Is that because of the order of the parameters changed? If not, why did the results change? Did we change seeds? The change is very low but still, I wouldn't expect the modifications to influence old results.

The regolding is due to the switch to `striginfyExact` in `SamplerFullSolveMultiApp`. Before, with `MultiAppSamplerControl`, we were only transferring 6 digits of precision, now we are doing all 16. Since I'm using this functionality with `SamplerParameterTransfer`, I needed to this change to pass those tests. I made very sure that all `MultiAppSamplerControl` passed without this change, and all `SamplerParameterTransfer` passed with it.

File: framework/include/multiapps/MultiApp.h
Comment (zachmprince) [reply framework/include/multiapps/MultiApp.h:583]: This was the easy way out to avoid having `setAppOutputFileBase` be public. I can possibly do a similar "key" strategy that we do to get write access in `ReporterData`. I will let @loganharbour comment on this.

File: framework/src/outputs/FileOutput.C
Comment (zachmprince) [reply framework/src/outputs/FileOutput.C:107]: This is git trying to be smarter than it is. I didn't actually move these, but for some reason git thinks I did when I put the new functions in.

File: modules/stochastic_tools/doc/content/syntax/ParameterStudy/index.md
Comment (zachmprince) [reply modules/stochastic_tools/doc/content/syntax/ParameterStudy/index.md:112]: Good catch. I use "output" so much as a noun, I forgot it can be verb. English is dumb.

File: modules/stochastic_tools/include/reporters/StochasticMatrix.h
Comment (zachmprince) [reply modules/stochastic_tools/include/reporters/StochasticMatrix.h:15]: The main reason I didn't make this suggested change is that I use `GlobalParams` to set the `sampler` arguement, and I didn't want similar people to have unexpected outputs. One other reason is that `StochasticReporter` is a little more general as it isn't required that declaring objects, like `SamplerReporterTransfer`, to have the same sampler as the one inputted. This is something I'm not to strongly opinionated on.

File: modules/stochastic_tools/src/actions/ParameterStudyAction.C
Comment (zachmprince) [reply modules/stochastic_tools/src/actions/ParameterStudyAction.C:538]: Yes, it is for vector/tensor parameters. It also allows for sampler entries to be assigned to multiple parameters. The changes in this PR allow `SamplerParameterTransfer` to use this syntax, so those no reason to check this in the walker.

File: modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C
Comment (zachmprince) [reply modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:213]: How would you suggest propagating the values? The return is only a single bool.

File: modules/stochastic_tools/src/reporters/StochasticMatrix.C
Comment (zachmprince) [reply modules/stochastic_tools/src/reporters/StochasticMatrix.C:52]: The only way this check won't pass is if the user did something wrong in `sampler_column_names`, so I didn't see much reason to protect this under the `isParamValid` flag.

File: modules/stochastic_tools/src/transfers/SamplerParameterTransfer.C
Comment (zachmprince) [reply modules/stochastic_tools/src/transfers/SamplerParameterTransfer.C:97]: I tried to cache the pointer previously. But in `normal` and `batch-reset` modes the pointer will change, so we would have to look for it again anyway. I would like to make this faster though. Any suggestions? I could maybe add a special approach for `batch-restore`.

**Comment by zachmprince** [comment]
Griffin patch [here](https://github.inl.gov/ncrc/griffin/pull/969) and isopod is now allowed to fail.

File: modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C
Comment (grmnptr) [reply modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:217]: Not sure if I put the question correctly, I was checking if I understood the implementation correctly. Not to question the necessity of the line :).

**Review by grmnptr — CHANGES_REQUESTED** [review]
Okay, I think this is the last review from my side. I like it already. However, I have to admit that I have a very localized understanding on the framework so I'll ask for a second opinion from either @GiudGiud or @loganharbour .

File: framework/src/multiapps/FullSolveMultiApp.C
Comment (zachmprince) [reply framework/src/multiapps/FullSolveMultiApp.C:144]: I'm guessing because when there is a console output, the convergence of the sub-app is obvious, so there is not need to repeat the message, unless we are in verbose mode.

File: modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C
Comment (zachmprince) [inline modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:212]: @loganharbour

File: framework/include/multiapps/MultiApp.h
Comment (loganharbour) [reply framework/include/multiapps/MultiApp.h]: We decided to just make the method public. I'm fine with this - I don't think it's overly concerning.

File: modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C
Comment (grmnptr) [inline modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:250]: very nice!

**Review by grmnptr — DISMISSED** [review]

**Comment by grmnptr** [comment]
I assume the Griffin failure is unrelated. Thanks for the addition Zach, very cool stuff!

**Review by grmnptr — APPROVED** [review]

**Pull request merged**
